### DISCUSSION TITLE
Some beginner katas are poorly worded

### DISCUSSION THREAD

**Discussion by spiff1023** [codewars/codewars.com#2607] [General]
Hi,
I'm relatively new to codewars and to coding (at least in C). While I can solve some katas relatively fast, I found that some easier ranked katas are poorly worded and it is hard to understand the exact outcome that the author wants. At least when you're not familiar with the programming language.
Here some examples:
1) When I read "return an integer" I understand this as writing a function that returns an integer. But sometimes the author wants me to save the integer at some location in the memory instead. 
2) In the problem I read that I have to perform some actions on an integer but in the already existing code the function does not only get an integer but an array and another integer without any explanation.
3) I have to do give out an array with a size that depends on the input array. Additionally the function gets a size_t variable without any explanation. Only by accident I find that the size_t variable actually tells me the length of my needed array. Otherwise the problem would have taken much more time. 

Why isn't there a kind of structure or checklist that every Kata needs to pass? Imho every input into the function should be explained and the wanted result (e.g. returning sth., store or change a memory value) should be stated explicitly.

Pls tell me if you had similar experiences as a beginner or if I'm just missing sth. obvious.
Ty

**Comment by hobovsky** [comment]
Your problems mostly are caused by coincidence of following factors:
 - Arrays in C are very primitive and when you pass an array between functions, you have to pass an additional information, in additional parameter/variable/whatever, about it's size. This is usually considered a well-known fact for every C programmer, but it's true that newbies might be not aware of it.
 - Contrary to C, arrays in higher-level languages are represented in a more convenient way, and every array always "knows" its size, so no additional variable is needed. As an effect, arrays can be passed between functions without any additional variable, and their size can be inspected by calling `len(array)`, `array.size()`, `array.length`, or something similar. It's not possible in C.
 - Descriptions of kata are sometimes either incomplete, or made generic and language-agnostic, because having language-specific details in descriptions sometimes causes problems. That's why words "array", "list", "sequence", etc. are sometimes meant with their more abstract meaning of "an ordered collection of elements, however it's represented in language of your choice", rather than a C-specific, or Javascript-specific, or Python-specific, or Java-specific thing.

Currently accepted practice for C kata on Codewars is that they are allowed to use generic, language-unaware descriptions, but C translations should at least explain requirements and parameters in the initial solution setup, which you see when entering the kata trainer, for example in a following way:

![image](https://user-images.githubusercontent.com/23709795/154532287-d57a12c9-ebdd-4d2d-9e0b-6ca59608db87.png)
![image](https://user-images.githubusercontent.com/23709795/154532356-570b05d3-1bf5-488e-995a-764d458a4389.png)
![image](https://user-images.githubusercontent.com/23709795/154532435-6fc13ef0-d802-4e4c-841c-7960e4dd492c.png)

As you see, kata use various techniques for arrays, because C gives many possibilities:
 - accepting a pointer to first element of array, and its size as a separate parameter.
 - returning an array by filling in a pre-allocated array passed as an input argument, usually accompanied by an output argument for a size.
 - returning a dynamically allocated array and it's size (via output parameter).
 - ... and some others...

In C, the word "return a value" can sometimes ambiguously be understood as "return it by filling in an output argument". And since these are very common patterns in C, users are expected to be aware of them.

Having said that, there are also old kata, or C translators who do not conform to the above recommendation, and create totally uninformative snippets:

![image](https://user-images.githubusercontent.com/23709795/154532811-e6b96980-cb5b-4882-9a6e-808392d73e54.png)

While a semi-experienced C programmer can deduct what these are so-called output arguments, for a newbie the meaning of `unsigned *index1, unsigned *index2` can be mysterious.

It _could_ be possible to make a requirement for every kata to have two descriptions: one for C, and another for languages which make work with arrays easier. But for now it was decided that this would be too tedious, kata would be difficult to maintain, and language-specific descriptions cause some additional problems. But I can see how i could make some things easier, especially for total beginners.

My recommendation for you would be:
 - Learn about different possible ways of passing arguments (especially arrays) into functions, and returning values (especially arrays) from functions. `return` is not the only possibility in C to return a value from a function ("return" in a broader meaning than "using a `return` statement"). Read up on so-called "output arguments" and how they can be used in C to return information from a function to its caller.
 - Whenever you are confused by a setup of C kata, raise an issue or ask a question in its discourse. I've been trying to promote some newbie-friendly authoring practices for C kata, and I have to admit they got mostly adapted, but maybe something is still not clear, or you encountered an old kata which was created before the guidelines were introduced.
 - When in doubt, come to Codewars Discord and visit it's `#beginners` or `#c` channel. C user base is not big, but it has some very active members.

Have fun, and good luck!

**Reply by spiff1023** [reply]
Wow! Thank You very much. I didn't expect such a detailed answer after such a short time. Seems like the main problem is me not being familiar with the subtleties and variety of C. I'll see how I get along with future katas and maybe I visit the Discord server at some point. Thank You!
